text/0000-dyn-error-generic-member-access.md

```diff
@@ -0,0 +1,527 @@
+`backtrace` function, for accessing a `Backtrace` of the state of the stack
+when an error was created. For all other forms of context relevant to an error
+report, the `Error` trait provides the `context` and `provide_context` functions.
+
+As an example of how to use this interface to construct an error report, let’s
+explore how one could implement an error reporting type. In this example, our
+error reporting type will retrieve the source code location where each error in
+the chain was created (if it captured a location) and render it as part of the
+chain of errors. Our end goal is to get an error report that looks something
+like this:
+
+```
+Error:
+    0: Failed to read instrs from ./path/to/instrs.json
+        at instrs.rs:42
+    1: No such file or directory (os error 2)
+```
+
+The first step is to define or use a type to represent a source location. In
+this example, we will define our own:
+
+```rust
+struct Location {
+    file: &'static str,
+    line: usize,
+}
+```
+
+Next, we need to gather the location when creating our error types.
+
+```rust
+struct ExampleError {
+    source: std::io::Error,
+    location: Location,
+    path: PathBuf,
+}
+
+impl fmt::Display for ExampleError {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(fmt, "Failed to read instrs from {}", path.display())
+    }
+}
+
+fn read_instrs(path: &Path) -> Result<String, ExampleError> {
+    std::fs::read_to_string(path).map_err(|source| {
+        ExampleError {
+            source,
+            path: path.to_owned(),
+            location: Location {
+                file: file!(),
+                line: line!(),
+            },
+        }
+    })
+}
+```
+
+Then, we need to implement the `Error` trait to expose these members to the error reporter.
+
+```rust
+impl std::error::Error for ExampleError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        Some(&self.source)
+    }
+
+    fn provide_context<'a>(&'a self, mut request: Pin<&mut Request<'a>>) {
+        request.provide::<Location>(&self.location);
+    }
+}
+```
+
+And, finally, we create an error reporter that prints the error and its source
+recursively, along with any location data that was gathered
+
+```rust
+struct ErrorReporter(Box<dyn Error + Send + Sync + 'static>);
+
+impl fmt::Debug for ErrorReporter {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut current_error = Some(self.0.as_ref());
+        let mut ind = 0;
+
+        while let Some(error) = current_error {
+            writeln!(fmt, "    {}: {}", ind, error)?;
+
+            if let Some(location) = error.context::<Location>() {
+                writeln!(fmt, "        at {}:{}", location.file, location.line)?;
+            }
+
+            ind += 1;
+            current_error = error.source();
+        }
+
+        Ok(())
+    }
+}
+```
+
+As you can see the `Error` trait provides the facilities needed to create error
```

Comment:
I think you could delete this sentence without losing your point, or maybe make it more specific, referencing the specific portions of the example that highlight this?

Reply:
I just feel like I need some sort of conclusion, otherwise the guide ends super suddenly and it feels bad. Coming up with a good conclusion is proving difficult :(

Reply:
What if you move the example output from above down here and make the "intended output" above a bit more generic?

```
Error:
    0: ERROR MESSAGE
        at LOCATION
    1: PRIOR MESSAGE
    ...
```